### ISSUE TITLE
use_self still gives false positives

### ISSUE THREAD

**Issue by imp** [rust-lang/rust-clippy#6902]
Now that #6818 is fixed and the fix (#6833) has made its way to the nightly there are still cases of false positives:

Lint name: `use_self`

I tried this code:

```rust
#![deny(clippy::use_self)]
use serde::Deserialize;

#[derive(Debug, Deserialize)]
#[serde(untagged)]
enum Direction {
    North,
}

fn main() {
    let d = Direction::North;

    println!("{:?}", d);
}
```

I expected to see this happen: clippy clean - no warning/errors

Instead, this happened:

```
    Checking playground v0.0.1 (/playground)
error: unnecessary structure name repetition
 --> src/main.rs:6:6
  |
6 |   enum Direction {
  |  ______^
7 | |     North,
  | |____^ help: use the applicable keyword: `Self`
  |
note: the lint level is defined here
 --> src/main.rs:1:9
  |
1 | #![deny(clippy::use_self)]
  |         ^^^^^^^^^^^^^^^^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#use_self

error: aborting due to previous error

error: could not compile `playground`

To learn more, run the command again with --verbose.
```

Interestingly enough it the code generated by `#[serde(untagged)]` that triggers this lint. If this line is commented out then clippy seems to be happy

### Meta

- `cargo clippy -V`: `clippy 0.1.52 (2021-03-13 acca818)`

- `rustc -Vv`:
 ```
rustc 1.52.0-nightly (acca81892 2021-03-13)
binary: rustc
commit-hash: acca818928654807ed3bc1ce0e97df118f8716c8
commit-date: 2021-03-13
host: x86_64-unknown-linux-gnu
release: 1.52.0-nightly
LLVM version: 12.0.0
 ```

**Comment by Y-Nak** [comment]
This issue would be automatically fixed when https://github.com/rust-lang/rust/issues/84122 is addressed.

**Comment by barafael** [comment]
For reference, here are two cases in which the same false positive occurs:
* [smart-default](https://docs.rs/smart-default/latest/smart_default/) with the `SmartDefault` derive macro
* [num-derive](https://docs.rs/num-derive/latest/num_derive/) with the `FromPrimitive` and `ToPrimitive` derive macros
    - This PR on `num-derive` proposes to stop using the type name and instead use `Self` in the generated code, solving the problem inside `num-derive`: https://github.com/rust-num/num-derive/pull/49

**Comment by carols10cents** [comment]
This problem has now made it to the clippy that's in beta (specifically `1.61.0-beta.1 (0f231250b 2022-04-05)`)

**Comment by kleinesfilmroellchen** [comment]
I can reproduce this as well. Shouldn't clippy not apply this lint (and others) on impls and functions marked as `#[automatically_derived]`?

**Comment by tyranron** [comment]
Now it fires on stable 😭

**Comment by repi** [comment]
The `use_self` lint is fantastic and we've run it multiple times manually to fix up our code with it, but even as of 1.63 these false positives on enums that use serde Serialize/Deserialize prevent us from enabling and enforcing use of the lint.

**Comment by daxpedda** [comment]
I believe this should now be fixable with the help of #8694.
